File: jarvis/internal/skills/system.go
package skills

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"
)

// ProcessSkill 进程管理技能
type ProcessSkill struct{}

// NewProcessSkill 创建进程管理技能
func NewProcessSkill() *ProcessSkill {
	return &ProcessSkill{}
}

// Name 技能名称
func (s *ProcessSkill) Name() string {
	return "process"
}

// Description 技能描述
func (s *ProcessSkill) Description() string {
	return "管理进程：启动、停止、查看状态"
}

// Parameters 参数定义
func (s *ProcessSkill) Parameters() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"action": map[string]interface{}{
				"type":        "string",
				"description": "操作类型",
				"enum":        []string{"list", "start", "stop", "find"},
			},
			"name": map[string]interface{}{
				"type":        "string",
				"description": "进程名称",
			},
			"pid": map[string]interface{}{
				"type":        "integer",
				"description": "进程ID",
			},
		},
		"required": []string{"action"},
	}
}

// Execute 执行进程操作
func (s *ProcessSkill) Execute(ctx context.Context, args json.RawMessage) (interface{}, error) {
	var params struct {
		Action string `json:"action"`
		Name   string `json:"name"`
		Pid    int    `json:"pid"`
	}
	if err := json.Unmarshal(args, &params); err != nil {
		return nil, fmt.Errorf("解析参数失败: %w", err)
	}

	switch params.Action {
	case "list":
		return s.listProcesses()
	case "find":
		if params.Name == "" {
			return nil, fmt.Errorf("需要指定进程名称")
		}
		return s.findProcess(params.Name)
	case "stop":
		if params.Pid == 0 {
			return nil, fmt.Errorf("需要指定进程ID")
		}
		return s.stopProcess(params.Pid)
	default:
		return nil, fmt.Errorf("不支持的操作: %s", params.Action)
	}
}

// listProcesses 列出进程
func (s *ProcessSkill) listProcesses() (interface{}, error) {
	// 使用ps命令列出进程
	cmd := exec.Command("ps", "aux")
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("获取进程列表失败: %w", err)
	}

	// 解析输出
	lines := strings.Split(string(output), "\n")
	processes := make([]map[string]string, 0)

	for i, line := range lines {
		if i == 0 || line == "" {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) >= 11 {
			processes = append(processes, map[string]string{
				"user":    fields[0],
				"pid":     fields[1],
				"cpu":     fields[2],
				"mem":     fields[3],
				"command": strings.Join(fields[10:], " "),
			})
		}
	}

	return map[string]interface{}{
		"success":   true,
		"processes": processes[:min(20, len(processes))], // 最多返回20个
		"total":     len(processes),
	}, nil
}

// findProcess 查找进程
func (s *ProcessSkill) findProcess(name string) (interface{}, error) {
	cmd := exec.Command("pgrep", "-l", name)
	output, err := cmd.Output()
	if err != nil {
		return map[string]interface{}{
			"success": true,
			"found":   false,
			"message": "未找到匹配的进程",
		}, nil
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	processes := make([]map[string]string, 0)

	for _, line := range lines {
		if line == "" {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) >= 2 {
			processes = append(processes, map[string]string{
				"pid":   fields[0],
				"name":  fields[1],
			})
		}
	}

	return map[string]interface{}{
		"success":   true,
		"found":     len(processes) > 0,
		"processes": processes,
	}, nil
}

// stopProcess 停止进程
func (s *ProcessSkill) stopProcess(pid int) (interface{}, error) {
	process, err := os.FindProcess(pid)
	if err != nil {
		return nil, fmt.Errorf("查找进程失败: %w", err)
	}

	if err := process.Signal(os.Interrupt); err != nil {
		return nil, fmt.Errorf("停止进程失败: %w", err)
	}

	return map[string]interface{}{
		"success": true,
		"message": fmt.Sprintf("已发送停止信号到进程 %d", pid),
		"pid":     pid,
	}, nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// DiskUsageSkill 磁盘使用技能
type DiskUsageSkill struct{}

// NewDiskUsageSkill 创建磁盘使用技能
func NewDiskUsageSkill() *DiskUsageSkill {
	return &DiskUsageSkill{}
}

// Name 技能名称
func (s *DiskUsageSkill) Name() string {
	return "disk_usage"
}

// Description 技能描述
func (s *DiskUsageSkill) Description() string {
	return "查看磁盘使用情况"
}

// Parameters 参数定义
func (s *DiskUsageSkill) Parameters() map[string]interface{} {
	return map[string]interface{}{
		"type":       "object",
		"properties": map[string]interface{}{},
	}
}

// Execute 执行查看磁盘
func (s *DiskUsageSkill) Execute(ctx context.Context, args json.RawMessage) (interface{}, error) {
	cmd := exec.Command("df", "-h")
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("获取磁盘信息失败: %w", err)
	}

	lines := strings.Split(string(output), "\n")
	disks := make([]map[string]string, 0)

	for i, line := range lines {
		if i == 0 || line == "" {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) >= 6 {
			disks = append(disks, map[string]string{
				"filesystem": fields[0],
				"size":       fields[1],
				"used":       fields[2],
				"available":  fields[3],
				"use_percent": fields[4],
				"mountpoint": fields[5],
			})
		}
	}

	return map[string]interface{}{
		"success": true,
		"disks":   disks,
	}, nil
}

// MemoryUsageSkill 内存使用技能
type MemoryUsageSkill struct{}

// NewMemoryUsageSkill 创建内存使用技能
func NewMemoryUsageSkill() *MemoryUsageSkill {
	return &MemoryUsageSkill{}
}

// Name 技能名称
func (s *MemoryUsageSkill) Name() string {
	return "memory_usage"
}

// Description 技能描述
func (s *MemoryUsageSkill) Description() string {
	return "查看内存使用情况"
}

// Parameters 参数定义
func (s *MemoryUsageSkill) Parameters() map[string]interface{} {
	return map[string]interface{}{
		"type":       "object",
		"properties": map[string]interface{}{},
	}
}

// Execute 执行查看内存
func (s *MemoryUsageSkill) Execute(ctx context.Context, args json.RawMessage) (interface{}, error) {
	// 读取/proc/meminfo
	data, err := os.ReadFile("/proc/meminfo")
	if err != nil {
		return nil, fmt.Errorf("读取内存信息失败: %w", err)
	}

	memInfo := make(map[string]string)
	lines := strings.Split(string(data), "\n")

	for _, line := range lines {
		if line == "" {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) >= 2 {
			key := strings.TrimSuffix(fields[0], ":")
			memInfo[key] = fields[1]
		}
	}

	return map[string]interface{}{
		"success":  true,
		"meminfo":  memInfo,
	}, nil
}

// ScheduleTaskSkill 定时任务技能
type ScheduleTaskSkill struct {
	tasks map[string]*scheduledTask
}

type scheduledTask struct {
	Name     string
	Command  string
	Interval time.Duration
	NextRun  time.Time
	Active   bool
}

// NewScheduleTaskSkill 创建定时任务技能
func NewScheduleTaskSkill() *ScheduleTaskSkill {
	return &ScheduleTaskSkill{
		tasks: make(map[string]*scheduledTask),
	}
}

// Name 技能名称
func (s *ScheduleTaskSkill) Name() string {
	return "schedule"
}

// Description 技能描述
func (s *ScheduleTaskSkill) Description() string {
	return "管理定时任务"
}

// Parameters 参数定义
func (s *ScheduleTaskSkill) Parameters() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"action": map[string]interface{}{
				"type":        "string",
				"description": "操作类型",
				"enum":        []string{"list", "add", "remove", "run"},
			},
			"name": map[string]interface{}{
				"type":        "string",
				"description": "任务名称",
			},
			"command": map[string]interface{}{
				"type":        "string",
				"description": "要执行的命令",
			},
			"interval": map[string]interface{}{
				"type":        "integer",
				"description": "执行间隔(秒)",
			},
		},
		"required": []string{"action"},
	}
}

// Execute 执行定时任务操作
func (s *ScheduleTaskSkill) Execute(ctx context.Context, args json.RawMessage) (interface{}, error) {
	var params struct {
		Action   string `json:"action"`
		Name     string `json:"name"`
		Command  string `json:"command"`
		Interval int    `json:"interval"`
	}
	if err := json.Unmarshal(args, &params); err != nil {
		return nil, fmt.Errorf("解析参数失败: %w", err)
	}

	switch params.Action {
	case "list":
		tasks := make([]map[string]interface{}, 0)
		for name, task := range s.tasks {
			tasks = append(tasks, map[string]interface{}{
				"name":     name,
				"command":  task.Command,
				"interval": task.Interval.Seconds(),
				"next_run": task.NextRun.Format(time.RFC3339),
				"active":   task.Active,
			})
		}
		return map[string]interface{}{
			"success": true,
			"tasks":   tasks,
		}, nil

	case "add":
		if params.Name == "" || params.Command == "" || params.Interval == 0 {
			return nil, fmt.Errorf("需要指定任务名称、命令和间隔")
		}
		s.tasks[params.Name] = &scheduledTask{
			Name:     params.Name,
			Command:  params.Command,
			Interval: time.Duration(params.Interval) * time.Second,
			NextRun:  time.Now().Add(time.Duration(params.Interval) * time.Second),
			Active:   true,
		}
		return map[string]interface{}{
			"success": true,
			"message": fmt.Sprintf("已添加定时任务: %s", params.Name),
		}, nil

	case "remove":
		if params.Name == "" {
			return nil, fmt.Errorf("需要指定任务名称")
		}
		delete(s.tasks, params.Name)
		return map[string]interface{}{
			"success": true,
			"message": fmt.Sprintf("已删除定时任务: %s", params.Name),
		}, nil

	case "run":
		if params.Name == "" {
			return nil, fmt.Errorf("需要指定任务名称")
		}
		task, ok := s.tasks[params.Name]
		if !ok {
			return nil, fmt.Errorf("任务不存在: %s", params.Name)
		}
		// 立即执行
		cmd := exec.Command("sh", "-c", task.Command)
		output, err := cmd.CombinedOutput()
		if err != nil {
			return map[string]interface{}{
				"success": false,
				"error":   err.Error(),
				"output":  string(output),
			}, nil
		}
		return map[string]interface{}{
			"success": true,
			"output":  string(output),
		}, nil

	default:
		return nil, fmt.Errorf("不支持的操作: %s", params.Action)
	}
}

File: nomad_mem_v4/nomad_mem/skills/system.py
"""
系统信息技能 - system_info, process, disk_usage, memory_usage
"""
import os
import platform
import psutil
from nomad_mem.skills.base import BaseSkill


class SystemInfo(BaseSkill):
    """获取系统信息"""

    @property
    def name(self) -> str:
        return "system_info"

    @property
    def description(self) -> str:
        return "获取当前系统的基本信息，包括操作系统版本、CPU、内存、主机名等。"

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {},
            "required": [],
        }

    def execute(self, args: dict) -> str:
        info = []
        info.append(f"操作系统: {platform.system()} {platform.release()}")
        info.append(f"主机名: {platform.node()}")
        info.append(f"架构: {platform.machine()}")
        info.append(f"Python: {platform.python_version()}")
        info.append(f"CPU核心: {os.cpu_count()}")

        mem = psutil.virtual_memory()
        info.append(f"内存总量: {mem.total / (1024**3):.1f} GB")
        info.append(f"内存可用: {mem.available / (1024**3):.1f} GB")

        return "\n".join(info)


class Process(BaseSkill):
    """进程管理"""

    @property
    def name(self) -> str:
        return "process"

    @property
    def description(self) -> str:
        return "管理进程：list列出所有进程，find查找特定进程，stop停止进程。"

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "action": {
                    "type": "string",
                    "description": "操作: list/查找/停止",
                    "enum": ["list", "find", "stop"],
                },
                "name": {
                    "type": "string",
                    "description": "进程名称（find/stop时需要）",
                },
                "limit": {
                    "type": "integer",
                    "description": "列出进程的最大数量，默认20",
                    "default": 20,
                },
            },
            "required": ["action"],
        }

    def execute(self, args: dict) -> str:
        action = args.get("action", "list")
        name = args.get("name", "")
        limit = args.get("limit", 20)

        try:
            if action == "list":
                processes = []
                for proc in psutil.process_iter(["pid", "name", "cpu_percent", "memory_percent"]):
                    try:
                        info = proc.info
                        processes.append(
                            f"PID: {info['pid']}, "
                            f"名称: {info['name']}, "
                            f"CPU: {info['cpu_percent'] or 0:.1f}%, "
                            f"内存: {info['memory_percent'] or 0:.1f}%"
                        )
                    except (psutil.NoSuchProcess, psutil.AccessDenied):
                        continue

                return f"前{limit}个进程:\n" + "\n".join(processes[:limit])

            elif action == "find":
                if not name:
                    return "find操作需要提供进程名称"

                results = []
                for proc in psutil.process_iter(["pid", "name"]):
                    try:
                        if name.lower() in proc.info["name"].lower():
                            results.append(f"PID: {proc.info['pid']}, 名称: {proc.info['name']}")
                    except (psutil.NoSuchProcess, psutil.AccessDenied):
                        continue

                if not results:
                    return f"未找到包含'{name}'的进程"
                return f"匹配的进程:\n" + "\n".join(results)

            elif action == "stop":
                if not name:
                    return "stop操作需要提供进程名称"

                stopped = []
                for proc in psutil.process_iter(["pid", "name"]):
                    try:
                        if name.lower() in proc.info["name"].lower():
                            proc.terminate()
                            stopped.append(proc.info["pid"])
                    except (psutil.NoSuchProcess, psutil.AccessDenied):
                        continue

                if not stopped:
                    return f"未找到包含'{name}'的进程"
                return f"已停止PID为 {', '.join(map(str, stopped))} 的进程"
            else:
                return f"未知操作: {action}"

        except Exception as e:
            return f"操作错误: {e}"


class DiskUsage(BaseSkill):
    """磁盘使用情况"""

    @property
    def name(self) -> str:
        return "disk_usage"

    @property
    def description(self) -> str:
        return "查看磁盘使用情况，包括总容量、已用空间、可用空间和百分比。"

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "查看路径，默认根目录",
                    "default": "/",
                },
            },
            "required": [],
        }

    def execute(self, args: dict) -> str:
        path = args.get("path", "/")

        try:
            usage = psutil.disk_usage(path)
            return (
                f"路径: {path}\n"
                f"总容量: {usage.total / (1024**3):.1f} GB\n"
                f"已使用: {usage.used / (1024**3):.1f} GB\n"
                f"可用: {usage.free / (1024**3):.1f} GB\n"
                f"使用率: {usage.percent:.1f}%"
            )
        except Exception as e:
            return f"获取磁盘信息错误: {e}"


class MemoryUsage(BaseSkill):
    """内存使用情况"""

    @property
    def name(self) -> str:
        return "memory_usage"

    @property
    def description(self) -> str:
        return "查看内存使用情况，包括总量、已用、可用、使用率等。"

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {},
            "required": [],
        }

    def execute(self, args: dict) -> str:
        try:
            mem = psutil.virtual_memory()
            swap = psutil.swap_memory()

            lines = [
                "=== 物理内存 ===",
                f"总量: {mem.total / (1024**3):.1f} GB",
                f"已使用: {mem.used / (1024**3):.1f} GB",
                f"可用: {mem.available / (1024**3):.1f} GB",
                f"使用率: {mem.percent:.1f}%",
                "",
                "=== 交换内存 ===",
                f"总量: {swap.total / (1024**3):.1f} GB",
                f"已使用: {swap.used / (1024**3):.1f} GB",
                f"可用: {swap.free / (1024**3):.1f} GB",
                f"使用率: {swap.percent:.1f}%",
            ]

            return "\n".join(lines)
        except Exception as e:
            return f"获取内存信息错误: {e}"
